src/v/datalake/record_multiplexer.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "container/fragmented_vector.h"
+#include "datalake/schemaless_translator.h"
+#include "json/writer.h"
+#include "model/record.h"
+
+// TODO Only needed for method definitions, not class. Move methods out of line
+// if possible
+#include "storage/parser_utils.h"
+
+#include <seastar/core/future.hh>
+
+#include <__concepts/same_as.h>
+
+#include <optional>
+#include <type_traits>
+
+namespace datalake {
+
+/*
+Consumes logs and sends records to the appropriate translator
+based on the schema ID. This is meant to be called with a
+read_committed_reader created from a kafka::partition_proxy.
+
+This is a skeleton class. Currently it only uses the trivial
+schemaless_translator.
+
+pseudocode:
+for each record {
+    t = get_translator(record.schema_id);
+    d = t.translate(record);
+    w = get_data_writer(record.schema_id);
+    w.record(d);
+}
+
+The WriterFactory is a class with
+- instance method `create_writer`
+- nested class `data_writer_result`
+- nested class `data_writer`
+  - instance method `add_data_struct` to consume a parsed record
+  - instance method `finish` to clean up and return a `data_writer_result`
+*/
+template<typename T>
+concept DataWriter = requires(
+  T a, schemaless_record_data& data, int64_t approx_size) {
+    { a.add_data_struct(data, approx_size) } -> std::same_as<bool>;
+    a.finish();
+};
+
+template<typename T>
+concept CreatesWriters = requires(T a, iceberg::struct_type schema) {
+    { a.create_writer(std::move(schema)) } -> DataWriter;
+};
+
+template<CreatesWriters WriterFactory>
+class record_multiplexer {
+public:
+    record_multiplexer(WriterFactory& writer_factory)
+      : _translator{schemaless_translator()}
+      , _writer_factory{writer_factory} {}
+    ss::future<ss::stop_iteration> operator()(model::record_batch&& batch);
+    ss::future<fragmented_vector<typename WriterFactory::data_writer_result>>
+    end_of_stream();
+
+private:
+    using translator = std::variant<schemaless_translator>;
+
+    translator& get_translator();
+    WriterFactory::data_writer& get_writer();
+
+    // TODO: in a future PR this will be a map of translators keyed by schema_id
+    translator _translator;
+    WriterFactory _writer_factory;
+
+    // TODO: similarly this will be a map keyed by schema_id
+    std::optional<typename WriterFactory::data_writer> _writer;
+};
+
+template<CreatesWriters WriterFactory>
```

Comment:
are we gaining anything from doing everything with templates, as opposed to creating some virtual interfaces? perhaps it would be simpler and we could avoid having so much stuff in the header and stuff like std::visit (or jumping through hoops to move bits out of the header)?

Reply:
The only thing we really gain is the ability to parameterize the result type, but I don't think that's terribly important. I can switch this to using an interface/virtual methods.